### PR TITLE
Spark discussion on RSpec's Implicit Block Expectation Syntax

### PR DESCRIPTION
**Pull request by pirj**

### CODE REVIEW

File: _posts/2019-07-17-rspec-implicit-block-syntax.markdown [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+```ruby
+subject { -> { do_something } }
+it { is_expected.to change(something).to(new_value) }
+```
+
+As you know, [`change`](https://relishapp.com/rspec/rspec-expectations/v/3-8/docs/built-in-matchers/change-matcher) is a block matcher.
+The above Implicit Block Expectation Syntax can be expressed with explicit syntax:
+
+```ruby
+it { expect { do_something }.to change(something).to(new_value) }
```
Comment (bbatsov) [inline _posts/2019-07-17-rspec-implicit-block-syntax.markdown]: Normally you'd use this with some description alongside `it`.

File: _posts/2019-07-17-rspec-implicit-block-syntax.markdown [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+author: Phil Pirozhkov
+---
+
+Hey, everyone!
+
+We planned to add a guideline to recommend against using an RSpec practice codenamed Implicit Block Expectation Syntax, but received some controversial feedback.
+In order to keep the [RSpec style guide](https://rspec.rubystyle.guide) reflecting the opinion of the Ruby community, we decided to highlight this topic and start a discussion.
+
+The syntax is as following:
+```ruby
```
Comment (bbatsov) [inline _posts/2019-07-17-rspec-implicit-block-syntax.markdown:16]: It's best to put have blank lines around code blocks.

File: _posts/2019-07-17-rspec-implicit-block-syntax.markdown [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+layout: post
+title:  "Implicit Block Syntax"
+date:   2019-07-17 10:15:00 +0300
+categories: rspec
+author: Phil Pirozhkov
+---
+
+Hey, everyone!
+
+We planned to add a guideline to recommend against using an RSpec practice codenamed Implicit Block Expectation Syntax, but received some controversial feedback.
```
Comment (bbatsov) [inline _posts/2019-07-17-rspec-implicit-block-syntax.markdown]: Probably you should put the name of the pattern in "" to stand out a bit more.

File: _posts/2019-07-17-rspec-implicit-block-syntax.markdown [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+As you know, [`change`](https://relishapp.com/rspec/rspec-expectations/v/3-8/docs/built-in-matchers/change-matcher) is a block matcher.
+The above Implicit Block Expectation Syntax can be expressed with explicit syntax:
+
+```ruby
+it 'changes something to a new value' do
+  expect { do_something }.to change(something).to(new_value) }
+end
+```
+
+Pros that we're aware of are:
```
Comment (bbatsov) [inline _posts/2019-07-17-rspec-implicit-block-syntax.markdown]: I guess you can specify which pros we are referring to here for clarity.

File: _posts/2019-07-17-rspec-implicit-block-syntax.markdown [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+ - no need to repeat `expect { ... }`
+
+Cons:
+ - an exception to the case when `subject` is cached, e.g. on each `is_subject` call the passed lambda will be re-executed
+ - rarely used and not unfamiliar to many people
+ - barely possible to detect it as a block expectation syntax with static analysis tools (RuboCop, Reek)
+ - the guide [doesn't recommend using one-liner syntax](https://rspec.rubystyle.guide/#use-subject) except in some cases
+ - there are no tests or documentation in RSpec that cover this syntax
+
+RSpec core team opinion on the topic:
```
Comment (bbatsov) [inline _posts/2019-07-17-rspec-implicit-block-syntax.markdown]: Here's ...

File: _posts/2019-07-17-rspec-implicit-block-syntax.markdown [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+[Jon Rowe](https://github.com/rspec/rspec-expectations/issues/805#issuecomment-112238533)
+
+The discussion is [here](https://github.com/rubocop-hq/rspec-style-guide/issues/76).[^1]
+
+In the spirit of broader community collaboration we've started this blog so that more people can see what we're working on, and participate in the discussions around introducing new guidelines and changing existing ones.[^2]
+
+Feel free to leave your reaction (a thumbs up or down, heart, rocket, or even a confused face) on the issue.
+Thumbs up will count towards recommending against this syntax, and thumbs down to leave out this guideline.
+
+We encourage you to bring additional arguments pro and con in the comments.
```
Comment (bbatsov) [inline _posts/2019-07-17-rspec-implicit-block-syntax.markdown]: pro and con -> for/against

File: _posts/2019-07-17-rspec-implicit-block-syntax.markdown [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+
+The expected outcome of the discussion can be either leaving the guideline out, or adding a guideline that would recommend against this syntax, adding a [`rubocop-rspec`](https://github.com/rubocop-hq/rubocop-rspec) cop to enforce it, and doing our best to make it practically impossible in the next major release of RSpec itself.
```
Comment (bbatsov) [inline _posts/2019-07-17-rspec-implicit-block-syntax.markdown]: recommend avoiding/suggest avoiding

File: _posts/2019-07-17-rspec-implicit-block-syntax.markdown [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+But it's all up to you how it will end up!
+
+We highly appreciate if you take time to glance over [the other open questions](https://github.com/rubocop-hq/rspec-style-guide/issues) and leave your reaction.
```
Comment (bbatsov) [inline _posts/2019-07-17-rspec-implicit-block-syntax.markdown]: take the time

File: _posts/2019-07-17-rspec-implicit-block-syntax.markdown [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+Basing on your activity, we'll pick what to discuss next, or will go ahead and add guidelines that get a prevailing number of positive reactions.
```
Comment (bbatsov) [inline _posts/2019-07-17-rspec-implicit-block-syntax.markdown]: get -> got

**Review by bbatsov — APPROVED** [review]
I'm totally for discouraging this confusing syntax. I personally hate it. :)

Most of my remarks are about small grammar issues.

**Comment by pirj** [comment]
@bbatsov, thanks for the review.

**Pull request merged**